docs/architecture/adr-064-abci-2.0.md

```diff
@@ -0,0 +1,443 @@
+height. Specifically, the proposer of the next block will receive the vote extensions
+in `RequestPrepareProposal.local_last_commit.votes`.
+
+If the application does not have vote extension information to provide, it
+returns a 0-length byte array as its vote extension.
+
+**NOTE**: 
+
+* Although each validator process submits its own vote extension, ONLY the *proposer*
+  of the *next* block will receive all the vote extensions included as part of the
+  pre-commit phase of the previous block. This means only the proposer will
+  implicitly have access to all the vote extensions, via `RequestPrepareProposal`,
+  and that not all vote extensions may be included, since a validator does not
+  have to wait for all pre-commits, only 2/3.
+* The pre-commit vote is signed independently from the vote extension.
+
+### `VerifyVoteExtension`
+
+This method allows validators to validate the vote extension data attached to
+each pre-commit message it receives. If the validation fails, the whole pre-commit
+message will be deemed invalid and ignored by CometBFT.
+
+CometBFT uses `VerifyVoteExtension` when validating a pre-commit vote. Specifically,
+for a pre-commit, CometBFT will:
+
+* Reject the message if it doesn't contain a signed vote AND a signed vote extension
+* Reject the message if the vote's signature OR the vote extension's signature fails to verify
+* Reject the message if `VerifyVoteExtension` was rejected by the app
+
+Otherwise, CometBFT will accept the pre-commit message.
+
+Note, this has important consequences on liveness, i.e., if vote extensions repeatedly
+cannot be verified by correct validators, CometBFT may not be able to finalize
+a block even if sufficiently many (+2/3) validators send pre-commit votes for
+that block. Thus, `VerifyVoteExtension` should be used with special care.
+
+CometBFT recommends that an application that detects an invalid vote extension
+SHOULD accept it in `ResponseVerifyVoteExtension` and ignore it in its own logic.
+
+### `FinalizeBlock`
+
+This method delivers a decided block to the application. The application must
+execute the transactions in the block deterministically and update its state
+accordingly. Cryptographic commitments to the block and transaction results,
+returned via the corresponding parameters in `ResponseFinalizeBlock`, are
+included in the header of the next block. CometBFT calls it when a new block
+is decided.
+
+In other words, `FinalizeBlock` encapsulates the current ABCI execution flow of
+`BeginBlock`, one or more `DeliverTx`, and `EndBlock` into a single ABCI method.
+CometBFT will no longer execute requests for these legacy methods and instead
+will just simply call `FinalizeBlock`.
+
+## Decision
+
+We will discuss changes to the Cosmos SDK to implement ABCI 2.0 in two distinct
+phases, `VoteExtensions` and `FinalizeBlock`.
+
+### `VoteExtensions`
+
+Similarly for `PrepareProposal` and `ProcessProposal`, we propose to introduce
```

Comment:
To clarify - are the following statements correct? If so, it would be nice to add some clarifying statements in the ADR

1. For each block round, every validator node will call `ExtendVoteHandler` 
2. For each block round and for each vote they receive from other nodes, every validator node will call `VerifyVoteExtensionHandler`

So if there are X number of validators for a block height H round 1, at the network level, there will be X number of calls to 1 and up to (X-1) * X calls to 2?

Reply:
Of course, happy to clarify!

1. Yes, for every round, CometBFT will execute an `ExtendVote` request. In fact, it does this for pretty much all ABCI methods.
2. Yes, this is also correct from my understanding. For each height and round within that height, CometBFT will execute a `VerifyVoteExtension` request.

Ideally, when you extend the vote and then verify all incoming vote exts, you receive 2/3+ valid ones and avoid having to proceed to the next round.

> So if there are X number of validators for a block height H round 1, at the network level, there will be X number of calls to 1 and up to (X-1) * X calls to 2?

Yes, I believe there will be X calls to `ExtendVote`. For `VerifyVoteExtension`, the proposer just needs 2/3 pre-commits to be valid, so this number can vary, but in theory it can be up to X-1, yes.

Reply:
Updated to include a statement on this. LMK if you think it suffices :)

Reply:
sg, thanks for clarifying. feel free to resolve this comment